p4c_bm/gen_json.py

```diff
@@ -1198,6 +1210,30 @@ def dump_field_aliases(json_dict, hlir, path_field_aliases):
     field_aliases = [[a, v] for a, v in aliases_dict.items()]
     json_dict["field_aliases"] = field_aliases
 
+def dump_extern_instances(json_dict, hlir):
+    extern_instances = []
+    id_ = 0
+    for name, p4_extern_instance in hlir.p4_extern_instances.items():
+        extern_instance_dict = OrderedDict()
+        extern_instance_dict["name"] = name
+        extern_instance_dict["id"] = id_
+        extern_instance_dict["type"] = p4_extern_instance.extern_type.name
+
+        id_ += 1
+
+        attributes = []
+        for attribute, attr in p4_extern_instance.attributes.items():
+            attribute_dict = OrderedDict()
+            attribute_dict["name"] = attribute
+            attribute_dict["type"] = "hexstr" # Ref: P4Objects.cpp, line 264
+            attribute_dict["value"] = attr.strip()
```

Comment:
Hi Antonin,

thanks for your comments. In my test I've used strings as attributes, but
always representing integer values You can have a look here:
https://github.com/engjefersonsantiago/p4-programs/blob/master/examples/rohc_compression/p4src/rohc_compression.p4
(searching the extern types).

About the travis report, I've seen in the log it is using the requirements,
not the requirements_v1.1. How can I execute the autochecks using the
correct HLIR version?

Thanks again and regards,
Jeferson

2016-09-12 16:23 GMT-04:00 Antonin Bas <notifications@github.com>:

> In p4c_bm/gen_json.py
> <https://github.com/p4lang/p4c-bm/pull/62#discussion_r78447245>:
>
> > +    extern_instances = []
> > +    id_ = 0
> > +    for name, p4_extern_instance in hlir.p4_extern_instances.items():
> > +        extern_instance_dict = OrderedDict()
> > +        extern_instance_dict["name"] = name
> > +        extern_instance_dict["id"] = id_
> > +        extern_instance_dict["type"] = p4_extern_instance.extern_type.name
> > +
> > +        id_ += 1
> > +
> > +        attributes = []
> > +        for attribute, attr in p4_extern_instance.attributes.items():
> > +            attribute_dict = OrderedDict()
> > +            attribute_dict["name"] = attribute
> > +            attribute_dict["type"] = "hexstr" # Ref: P4Objects.cpp, line 264
> > +            attribute_dict["value"] = attr.strip()
>
> As you correctly noted, only hexstr attributes are supported in bmv2
> (currently). As a result you should probably check that attr has the
> right type and issue a LOG_CRITICAL if it is not. Btw, you seem to assume
> that attr is a string; but I would believe an integer value would make
> more sense. hexstr is supposed to represent an integer value. What kind
> of attribute type did you use in your P4 programs; did you use string? I
> think that given the current extent of extern support in bmv2, it would
> make more sense to use integer attributes (bit or int type?).
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/p4lang/p4c-bm/pull/62/files/cab57c92e42fa8f56b496094495bf348fb824a37#r78447245>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AKNH5jEQLHhDFsIpmWTly1XUe7mSby2Xks5qpbTOgaJpZM4J5GHs>
> .
>



-- 
Jeferson Santiago da Silva, M.Sc.
<http://br.linkedin.com/in/engjefersonsantiago>
Polytechnique Montréal
Ph.D. Candidate in Computer Engineering
Hardware and Programmable Logic Design Engineer
Phone: +1 (514) 581-1004
Skype: eng.jefersonsantiago
Montreal/QC - Canada
Lattes CV: http://lattes.cnpq.br/9568850258686622